### ISSUE TITLE
Week days and week count

### ISSUE THREAD

**Issue by talamaska** [jama5262/jiffy#58]
I'm not sure why is this implemented that way, for me this is incorrect, everybody knows a week starts from Monday, but instead your week day is 0-6 Sunday to Saturday?
Next big thing with the same logic, one record is on Saturday, next one is on Sunday and your logic makes them in different weeks. This is wrong. Maybe consider having a setting to set which is the first day of the week so that the calculations are correct. This is a show stopper for me. I'm trying to group records by week to show a bar chart, but I'm getting wrong results.

**Comment by schankam** [comment]
@talamaska This is quite standard when working with dates, in Javascript it is exactly behaving the same ways. Weeks start on Sunday and end on Saturday; They also start with Sunday = 0 and Saturday = 6; Same goes for Month, they start at 0 and end at 11.

**Comment by talamaska** [comment]
Then why in the original DateTime in Dart Monday is 1 and Sunday is 7. 
Why not make that consistent with the Dart API instead of implementing something from Javascript which is different than in Dart, this is confusing to say at least. 
https://api.dart.dev/stable/2.7.2/dart-core/DateTime/weekday.html

**Comment by andzejsw** [comment]
Is there some way, to change week starting day?

**Comment by schankam** [comment]
@talamaska I agree with you and at first I was basing my comment on what I saw on the documentation, but what you're saying indeed makes more sens.

**Comment by jama5262** [comment]
@talamaska @schankam I understand it was my mistake on my side. That the numbering of the weekdays is not well implemented based on Dart. That is the numbering should be from 1 -7 and not 0 - 6

But on the question on which day to start, it was based on locale and I chose the most used that is `Sunday to Saturday`. And this lead to other locales not getting the correct start day. It was later updated, a few months ago, where the start date would change based on the locale you chose e.g

`await Jiffy.locale("fr"); // the start date would be Sunday`
`await Jiffy.locale("ar"); // the start date would be Saturday`

But all in all, we should implement as @talamaska suggested a setting for this, on which you can choose which start day you want

**Comment by talamaska** [comment]
ok, that information was missing from the docs. I didn't know that it depends on the locale. But still, it's kind of strange when you make some comparisons, in one locale to show that they are on the same week, and on the other to show that they are not. Thanks for explaining that behavior.

**Comment by pratikpparikh** [comment]
@jama5262 the solution of using locale to adjust for date is very poor solution.  First of it return date that is in that locale after that point on wards and that is not always required or needed.  Why can't there be a constructor that accepts a starting week day as an param which will then be used in the implementation if provided.

**Comment by Adrian4777** [comment]
Does anyone know if this is implemented yet? Jiffy().week is giving me the wrong week, I would like to use Monday-Sunday.

**Comment by akeblom** [comment]
> Does anyone know if this is implemented yet? Jiffy().week is giving me the wrong week, I would like to use Monday-Sunday.

Me too! Please implement!

**Comment by Adrian4777** [comment]
```
/// Calculates week number from a date as per https://en.wikipedia.org/wiki/ISO_week_date#Calculation
int weekNumber(DateTime date) {
  int dayOfYear = int.parse(DateFormat("D").format(date));
  int week = ((dayOfYear - date.weekday + 10) / 7).floor();
  if (week == 0) {
    return 53;
  } else {
    return week;
  }
}
```

**Comment by akeblom** [comment]
> @akeblom, try this:
> 
> ```
> /// Calculates week number from a date as per https://en.wikipedia.org/wiki/ISO_week_date#Calculation
> int weekNumber(DateTime date) {
>   int dayOfYear = int.parse(DateFormat("D").format(date));
>   int week = ((dayOfYear - date.weekday + 10) / 7).floor();
>   if (week == 0) {
>     return 53;
>   } else {
>     return week;
>   }
> }
> ```

Thanks!

**Comment by SuperKrallan** [comment]
@schankam @talamaska : The start of the week is locale dependent. I.e. in the US the week starts on a Sunday, but in Sweden the week starts on a Monday. Since I live in Sweden and have 'sv' as locale, I would like to see the Jiffy(date).startOf(Units.WEEK).dateTime give me the date of the Monday for that week. Now I get Sunday...

**Comment by SuperKrallan** [comment]
...and I also found and had to implement Adrian4777s solution above for the week number. And in this context I also found the following code snippet that I liked and implemented:

/// Calculates number of weeks for a given year as per https://en.wikipedia.org/wiki/ISO_week_date#Weeks_per_year
int numOfWeeks(int year) {
  DateTime dec28 = DateTime(year, 12, 28);
  int dayOfDec28 = int.parse(DateFormat("D").format(dec28));
  return ((dayOfDec28 - dec28.weekday + 10) / 7).floor();
}

**Comment by danielmahon** [comment]
It would be very helpful to set the locale and start of the week to the specific instance. We would use this for calculating dates based on work/school weeks where we want the weekend to include Saturday/Sunday. This would override the locale for that usage. The global setting should remain the default for the locale.

**Comment by jama5262** [comment]
Fixed in #294
